### DISCUSSION TITLE
Erro de autenticação com JWT

### DISCUSSION THREAD

**Discussion by RochaRaphael** [balta-io/forum#282] [Tira Dúvidas] [answered]
**Olá Balta, tudo bem? Estou com um problema para fazer a autenticação, a minha API parece estar gerando os tokens corretamente(verifiquei no jwt.io) mas quando vou chamar um método que precisa de autenticação, ele da o seguinte erro:**

_System.InvalidOperationException: No authenticationScheme was specified, and there was no DefaultChallengeScheme found. The default schemes can be set using either AddAuthentication(string defaultScheme) or AddAuthentication(Action<AuthenticationOptions> configureOptions).
   at Microsoft.AspNetCore.Authentication.AuthenticationService.ChallengeAsync(HttpContext context, String scheme, AuthenticationProperties properties)
   at Microsoft.AspNetCore.Authorization.Policy.AuthorizationMiddlewareResultHandler.HandleAsync(RequestDelegate next, HttpContext context, AuthorizationPolicy policy, PolicyAuthorizationResult authorizeResult)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)_
   
   
   Pesquisando, entendi que o mesmo só acontece quando ocorre uma tentativa de autenticação sem ter definido a política de autenticação corretamente.
   
   Porém revisei o código várias vezes, tentei refaze-lo e o mesmo erro persiste.

   OBS: Instalei os pacotes 
 ```
  - dotnet add package Microsoft.AspNetCore.Authentication
   - dotnet add package Microsoft.AspNetCore.Authentication.JwtBearer --version 6
```
   A versão 6 é a compatível.
   
   Vou deixar aqui as partes do meu código referente ao jwt:
   
### Program:
    var builder = WebApplication.CreateBuilder(args);
    ConfigureMvc(builder);
    ConfigureServices(builder);
    
    builder.Services.AddEndpointsApiExplorer();
    
    var app = builder.Build();
    LoadConfiguration(app);
    
    app.UseAuthentication();
    app.UseAuthorization();
    app.MapControllers();
    
    app.Run();
    
    void LoadConfiguration(WebApplication app)
    {
        Configuration.JwtKey = app.Configuration.GetValue<string>("JwtKey");
    }
    
    void ConfigureAuthentication(WebApplicationBuilder builder)
        {
            var key = Encoding.ASCII.GetBytes(Configuration.JwtKey);
            builder.Services.AddAuthentication(x =>
            {
                x.DefaultAuthenticateScheme = JwtBearerDefaults.AuthenticationScheme;
                x.DefaultChallengeScheme = JwtBearerDefaults.AuthenticationScheme;
            }).AddJwtBearer(x =>
            {
                x.TokenValidationParameters = new TokenValidationParameters
                {
                    ValidateIssuerSigningKey = true,
                    IssuerSigningKey = new SymmetricSecurityKey(key),
                    ValidateIssuer = false,
                    ValidateAudience = false
                };
            });
    }
    void ConfigureServices(WebApplicationBuilder builder)
    {
        var connectionString = builder.Configuration.GetConnectionString("DefaultConnection");
        builder.Services.AddDbContext<DataContext>(
            options =>
                options.UseSqlServer(connectionString));
        builder.Services.AddTransient<TokenService>();
    }

### Configuration:
    public class Configuration
        {
            public static string JwtKey = "ZmVkYWY3ZDg4NjNiNDhlMTk3YjkyODdkNDkyYjcwOGU=";
        }
        
    
### RoleClaimsExtension:
    public static class RoleClaimsExtension
        {
            public static IEnumerable<Claim> GetClaims(this User user)
            {
                var result = new List<Claim>
            {
                new(ClaimTypes.Name, user.Email)
            };
                result.AddRange(
                    user.Roles.Select(role => new Claim(ClaimTypes.Role, role.Slug))
                );
                return result;
            }
        }

### TokenService:
    public class TokenService
        {
            public string GenerateToken(User user)
            {
                var tokenHandler = new JwtSecurityTokenHandler();
                var key = Encoding.ASCII.GetBytes(Configuration.JwtKey);
                var claims = user.GetClaims();
                var tokenDescriptor = new SecurityTokenDescriptor
                {
                    Subject = new ClaimsIdentity(claims),
                    Expires = DateTime.UtcNow.AddHours(4),
                    SigningCredentials = new SigningCredentials(
                        new SymmetricSecurityKey(key),
                        SecurityAlgorithms.HmacSha256Signature)
                };
                var token = tokenHandler.CreateToken(tokenDescriptor);
                return tokenHandler.WriteToken(token);
            }
        }
    
    
### UserController:
      [Authorize]
              [HttpGet("v1/accounts/{id:int}")]
              public async Task<IActionResult> FindUserById(
                  [FromRoute] int id
                  )
              {
                  if (!ModelState.IsValid)
                      return BadRequest(new ResultViewModel<string>(ModelState.GetErrors()));

            try
            {
                var user = await userServices.GetByIdAsync(id);
                if (user == null)
                    return NotFound("User not found");

                return Ok(new ResultViewModel<ShowUserViewModel>(user));
            }
            catch
            {
                return StatusCode(500, new ResultViewModel<string>("42X00 - Server failure"));
            }
        }

**Caso queira ver o repositório completo do projeto, vou deixar o link aqui:** https://github.com/RochaRaphael/API-Ecommerce

**Comment by andrebaltieri** [comment, marked as answer]
Boa noite @RochaRaphael, como vai?
Acredito que faltou chamar o método `ConfigureAuthentication` que você criou, não foi?

**Reply by RochaRaphael** [reply]
Sim foi isso mesmo, muito obrigado Balta!

**Reply by andrebaltieri** [reply]
🚀
